h/services/bulk_annotation.py

```diff
@@ -176,7 +187,14 @@ def _audience_groups_subquery(cls, authority, audience):
             .join(GroupMembership, GroupMembership.group_id == Group.id)
             .join(cls._AUDIENCE, GroupMembership.user_id == cls._AUDIENCE.id)
             .where(
-                cls._AUDIENCE._username.in_(  # pylint:disable=protected-access
+                sa.func.lower(
```

Comment:
Use the index in username, which is defined with this functions, not the raw value.

Reply:
I'm not sure I understand this change? It looks like previously the code was not removing `.`'s from `cls._AUDIENCE._username` but now it is. Doesn't that change the results of the query?

Reply:
Sorry, I only mentioned the part about the index. My understanding is that:

- We are doing the `replace(".", "")` on the python side, the RHS of the condition on the query

- We were not doing that change on the LHS of the condition, but the the values themselves in the DB might have the `.` in the name


- I reckon, because of this, the index on `username` is not on the raw value but in the function replacing the dot. All queries to username must make the change in both side of the comparison.

- This might not matter that much anyway because LMS usernames don't have dots on them:


[Metabase query for LMS usernames](https://metabase.hypothes.is/question#eyJkYXRhc2V0X3F1ZXJ5Ijp7InR5cGUiOiJuYXRpdmUiLCJuYXRpdmUiOnsicXVlcnkiOiJzZWxlY3QgdXNlcm5hbWUgZnJvbSBcInVzZXJcIiB3aGVyZSBhdXRob3JpdHkgPSAnbG1zLmh5cG90aGVzLmlzJyBhbmQgcmVwbGFjZShcInVzZXJuYW1lXCIsICcuJywgJycpIDw-IHVzZXJuYW1lXG5cblxuIiwidGVtcGxhdGUtdGFncyI6e319LCJkYXRhYmFzZSI6NX0sImRpc3BsYXkiOiJ0YWJsZSIsInBhcmFtZXRlcnMiOltdLCJ2aXN1YWxpemF0aW9uX3NldHRpbmdzIjp7fX0=)


other usernames [in other authorities do have dots](https://metabase.hypothes.is/question#eyJkYXRhc2V0X3F1ZXJ5Ijp7InR5cGUiOiJuYXRpdmUiLCJuYXRpdmUiOnsicXVlcnkiOiJzZWxlY3QgdXNlcm5hbWUgZnJvbSBcInVzZXJcIiB3aGVyZSBhdXRob3JpdHkgPD4gJ2xtcy5oeXBvdGhlcy5pcycgYW5kIHJlcGxhY2UoXCJ1c2VybmFtZVwiLCAnLicsICcnKSA8PiB1c2VybmFtZSBsaW1pdCAxMDAiLCJ0ZW1wbGF0ZS10YWdzIjp7fX0sImRhdGFiYXNlIjo1fSwiZGlzcGxheSI6InRhYmxlIiwicGFyYW1ldGVycyI6W10sInZpc3VhbGl6YXRpb25fc2V0dGluZ3MiOnt9fQ==)


- So if my understanding is correct with this change we fix correctness for non-LMS users (we compare no-dot-usernames with themselves) and benefit from the index in the LMS context.


- I've no idea why we do all of this `.` business. I reckon it might have to do with elastic search?

Reply:
So is this query currently broken on `main`? If the bulk annotation API were used with non-LMS usernames that have `.`'s in them then this query would fail to find those?

Reply:
Yes, moved this to it's own PR 

https://github.com/hypothesis/h/pull/8178